tools/adaparse.cpp

```diff
@@ -1,8 +1,169 @@
 #include <cstdlib>
+#include <chrono>
 #include <iostream>
 #include <string_view>
+#include <sys/stat.h>
+#include <fmt/core.h>
 #include <cxxopts.hpp>
+#include <fstream>
+#include <fmt/os.h>
+#include <cerrno>
+
+#include <fmt/os.h>
+#include <typeinfo>
+#include <iostream>
+#include <cxxabi.h>
+#include <memory>
+
 #include "ada.h"
+#include "line_iterator.h"
+#ifdef _MSC_VER
+#include <io.h>
+#else
+#include <unistd.h>
+#endif
+
+uint64_t nano() {
+  return std::chrono::duration_cast<::std::chrono::nanoseconds>(
+             std::chrono::steady_clock::now().time_since_epoch())
+      .count();
+}
+
+template <typename Callable>
+bool print_part(Callable&& fmt_or_adaparse_print, std::string_view get_part,
+                const ada::url_aggregator& url) {
+  if (get_part.size() == 4) {
+    if (get_part[0] == 'h') {
+      if (get_part == "host") {
+        fmt_or_adaparse_print("{}\n", url.get_host());
+        return true;
+      } else if (get_part == "hash") {
+        fmt_or_adaparse_print("{}\n", url.get_hash());
+        return true;
+      }
+    } else if (get_part[0] == 'p') {
+      if (get_part == "port") {
+        fmt_or_adaparse_print("{}\n", url.get_port());
+        return true;
+      }
+    }
+  } else if (get_part.size() == 6) {
+    if (get_part == "origin") {
+      fmt_or_adaparse_print("{}\n", url.get_origin());
+      return true;
+    }
+    if (get_part == "search") {
+      fmt_or_adaparse_print("{}\n", url.get_search());
+      return true;
+    }
+  } else if (get_part.size() == 8) {
+    if (get_part[0] == 'p') {
+      if (get_part == "protocol") {
+        fmt_or_adaparse_print("{}\n", url.get_protocol());
+        return true;
+      }
+      if (get_part == "password") {
+        fmt_or_adaparse_print("{}\n", url.get_password());
+        return true;
+      }
+      if (get_part == "pathname") {
+        fmt_or_adaparse_print("{}\n", url.get_pathname());
+        return true;
+      }
+    }
+  } else if (get_part == "hostname") {
+    fmt_or_adaparse_print("{}\n", url.get_hostname());
+    return true;
+
+  } else if (get_part == "username") {
+    fmt_or_adaparse_print("{}\n", url.get_username());
+    return true;
+  }
+
+  fmt::print(stderr, "\"{}\" not found\n", get_part);
+  return false;
+}
+
+// This function parses a FILE * descriptor, line by line (URL by URL).It
+// applies the given Callable (in this case the lambda adaparse_print in the
+// main() function) to print the output either to a file on disk or to the
+// console, depending on arguments given.
+// It also optionally will output a benchmark to the console.
+template <typename Callable>
+int piped_file(Callable&& adaparse_print, const cxxopts::ParseResult result,
+               FILE* input_file) {
+  constexpr size_t cache_length = 32768;
+  std::unique_ptr<char[]> cachebuffer(new char[cache_length]{});
+
+  uint64_t before = nano();
+
+  size_t howmany = 0;
+  size_t tr;
```

Comment:
Can you select descriptive names for variables, or add comments on top of it? `tr` or `howmany` does not help me understand the context without checking all the usages.